ttnn/cpp/ttnn/operations/pool/generic/pool2d_utils.cpp

```diff
@@ -0,0 +1,280 @@
+        channels,
+        output_shape[1],
+        output_shape[2],
+        channels,
+        compute_grid_size,
+        ShardOrientation::ROW_MAJOR,
+        false,
+        false);
+    auto output_parallel_config_width =
+        conv::determine_output_parallel_config(input_parallel_config_width, compute_grid_size, channels, false);
+
+    auto input_parallel_config_block = conv::determine_parallel_config(
+        TensorMemoryLayout::BLOCK_SHARDED,
+        batch_size,
+        channels,
+        output_shape[1],
+        output_shape[2],
+        channels,
+        compute_grid_size,
+        ShardOrientation::COL_MAJOR,
+        false,
+        false);
+
+    auto get_memconfig = [&](const ParallelConfig& parallel_config) {
+        uint32_t nhw = batch_size * output_shape[1] * output_shape[2];
+        uint32_t out_channeal_padded = tt::round_up(
+            channels, conv::get_num_cores_channels_from_parallel_config(parallel_config) * tt::constants::TILE_WIDTH);
+        return conv::create_sharded_memory_config_from_parallel_config(
+            ttnn::Shape({1, 1, nhw, out_channeal_padded}), parallel_config, tt::constants::TILE_HEIGHT);
+    };
+
+    uint32_t l1_usage_height = calculate_L1_usage(
+        input_tensor,
+        sliding_window_config.window_hw.first,
+        sliding_window_config.window_hw.second,
+        sliding_window_config.get_output_shape()[1],
+        sliding_window_config.get_output_shape()[2],
+        get_memconfig(input_parallel_config_height),
+        get_memconfig(input_parallel_config_height));
+
+    uint32_t l1_usage_width = calculate_L1_usage(
+        input_tensor,
+        sliding_window_config.window_hw.first,
+        sliding_window_config.window_hw.second,
+        sliding_window_config.get_output_shape()[1],
+        sliding_window_config.get_output_shape()[2],
+        get_memconfig(input_parallel_config_width),
+        get_memconfig(output_parallel_config_width));
+
+    uint32_t l1_usage_block = calculate_L1_usage(
+        input_tensor,
+        sliding_window_config.window_hw.first,
+        sliding_window_config.window_hw.second,
+        sliding_window_config.get_output_shape()[1],
+        sliding_window_config.get_output_shape()[2],
+        get_memconfig(input_parallel_config_block),
+        get_memconfig(input_parallel_config_block));
+
+    uint32_t ncores_height = input_parallel_config_height.grid.num_cores();
+    uint32_t ncores_width = input_parallel_config_width.grid.num_cores();
+    uint32_t ncores_block = input_parallel_config_block.grid.num_cores();
+
+    uint32_t winning_l1_usage = l1_usage_height;
+    auto winning_config = input_parallel_config_height;
+    // Make sure that BS not only has smaller size but provides at least some slicing along the channels.
+    // In case we have BS that would slice the tensor only along the HS conv2d code would fail later on.
+    if (l1_usage_block < l1_usage_height && ncores_block > compute_grid_size.x) {
```

Comment:
I guess this logic for BS is c/p from convs I'm not sure if it's relevant for pools?